### PR TITLE
Split out environment discovery/load from running

### PR DESCRIPTION
**Pull request by charleso**
/cc @nhibberd @markhibberd @EduardSergeev 

Because I'm lazy it depends on https://github.com/ambiata/mismi/pull/184

### CODE REVIEW

File: mismi-s3/main/s3.hs [hunk 1]
```diff
@@ -88,9 +88,8 @@ run = \case
 
 runK :: AmazonkaCommand -> IO ()
 runK k = do
-  mr <- getRegionFromEnv
-  let r = fromMaybe Sydney mr
-  orDie renderError . EitherT . tryJust exceptExit . runAWSWithRegion r $ case k of
+  e <- orDie envErrorRender . EitherT $ discoverAwsEnv
```
Comment (charleso) [inline mismi-s3/main/s3.hs]: I'm open to better combinators for this top level consumption

Comment (markhibberd) [inline mismi-s3/main/s3.hs]: I think this plays into the conversation about having a better/more-consistent env-var based config library.

**Comment by charleso** [comment]
@nhibberd @EduardSergeev This look OK to you?

File: mismi-core/src/Mismi/Environment.hs [hunk 2]
```diff
@@ -40,5 +49,13 @@ getRegionFromEnv = do
     (either (throwM . RegionUnknown . T.pack) return . fromText . T.pack)
     mr
 
+discoverAwsEnv :: IO (Either EnvError Env)
+discoverAwsEnv = runEitherT $ do
+  r <- EitherT . fmap (maybeToRight MissingRegion) $ getRegionFromEnv
+  lift $ newEnv r Discover
```
Comment (nhibberd) [inline mismi-core/src/Mismi/Environment.hs]: Not sure if this was on purpose, but you have omitted the `AWS_SECURITY_TOKEN` [lookup](https://github.com/ambiata/mismi/pull/188/files#diff-4bf4d01490c6909528be1107308d244fL76). Do we not want to expose this anymore? (this was used to `sts` operations with these runners previously)

File: mismi-core/test/Test/Mismi.hs [hunk 3]
```diff
@@ -12,11 +13,20 @@ import           Mismi
...
 
 
-testAWS :: Testable a => Region -> AWS a -> Property
-testAWS r a = testIO $
-  (runAWSWithRegion r a >> return succeeded) `catch` (return . exception "AWS action failed with exception")
+testAWS :: Testable a => AWS a -> Property
+testAWS a = testIO $
+  (runAWSDefaultRegion a >> return succeeded) `catch` (return . exception "AWS action failed with exception")
+
+-- Default to Sydney for tests only, production should fail without the environment variable
```
Comment (nhibberd) [inline mismi-core/test/Test/Mismi.hs]: :+1:

File: mismi-core/src/Mismi/Environment.hs (hunk 2 above)
Comment (nhibberd) [inline mismi-core/src/Mismi/Environment.hs]: Sorry you can ignore the above comment. I see amazonka now supports this with `AWS_SESSION_TOKEN`

**Comment by nhibberd** [comment]
When builds are :green_book:

File: mismi-core/src/Mismi/Environment.hs [hunk 4]
```diff
@@ -6,11 +6,15 @@ module Mismi.Environment (
   , RegionError (..)
   , getRegionFromEnv
   , regionErrorRender
+  , discoverAwsEnv
```
Comment (EduardSergeev) [inline mismi-core/src/Mismi/Environment.hs]: Shouldn't it be `discoverAWSEnv` for consistency?

File: bin/ci.common [hunk 5]
```diff
@@ -14,6 +14,5 @@ for x in $MODULES; do
         cd $(dirname $x)
         ./cabal update || exit $?
         ./cabal build || exit $?
-        ./cabal test  --log=/dev/stdout || exit $?
```
Comment (EduardSergeev) [inline bin/ci.common]: Doesn't it mean tests don't get run anymore?

Comment (nhibberd) [inline bin/ci.common]: :boom:

Comment (jystic) [inline bin/ci.common]: do it live

**Comment by charleso** [comment]
Builds are green. Renamed `discoverAWSEnv`. Going to merge first thing after lunch.

**Comment by EduardSergeev** [comment]
:+1:

**Pull request merged**